From 1a0994a1e273a086d1169f6836404fc4be8a28fe Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Mon, 14 Apr 2025 19:51:40 +0200
Subject: [PATCH] Use GApplication to parse command line options

It already supports adding and parsing option entries, so it's not
necessary to parse them manually.
---
 src/gyahtzee.c | 66 +++++++++++++++++++++++---------------------------
 1 file changed, 30 insertions(+), 36 deletions(-)

diff --git a/src/gyahtzee.c b/src/gyahtzee.c
index 03be2b2..8b3c8e9 100644
--- a/src/gyahtzee.c
+++ b/src/gyahtzee.c
@@ -956,42 +956,11 @@ GyahtzeeActivateGame (GApplication *app, gpointer user_data)
   gtk_window_present (GTK_WINDOW (window));
 }
 
-int
-main (int argc, char *argv[])
+static int
+GyahtzeeLocalOptions (GApplication *app,
+                      GVariantDict *options,
+                      gpointer      user_data)
 {
-  GOptionContext *context;
-  gboolean retval;
-  GError *error = NULL;
-
-  setlocale (LC_ALL, "");
-  bindtextdomain (GETTEXT_PACKAGE, LOCALEDIR);
-  bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
-  textdomain (GETTEXT_PACKAGE);
-
-  application = gtk_application_new ("org.gnome.Tali", 0);
-  g_signal_connect (application, "startup", G_CALLBACK (GyahtzeeStartup), NULL);
-  g_signal_connect (application, "activate", G_CALLBACK (GyahtzeeActivateGame), NULL);
-
-  /* Reset all yahtzee variables before parsing args */
-  YahtzeeInit ();
-
-  context = g_option_context_new (NULL);
-  g_option_context_add_group (context, gtk_get_option_group (TRUE));
-  g_option_context_add_main_entries (context, yahtzee_options,
-				     GETTEXT_PACKAGE);
-  retval = g_option_context_parse (context, &argc, &argv, &error);
-  g_option_context_free (context);
-  if (!retval) {
-    g_print ("%s", error->message);
-    g_error_free (error);
-    exit (1);
-  }
-
-  settings = g_settings_new ("org.gnome.Tali");
-
-  g_set_application_name (_(appName));
-  g_set_prgname ("org.gnome.Tali");
-
   /* If we're in computer test mode, just run some tests, no GUI */
   if (test_computer_play > 0) {
       gint ii, jj, kk;
@@ -1028,9 +997,34 @@ main (int argc, char *argv[])
           }
       }
       printf ("Computer average: %.2f for %d trials\n", sum_scores / test_computer_play, NUM_TRIALS);
-      exit (0);
+      return 0;
   }
 
+  return -1;
+}
+
+int
+main (int argc, char *argv[])
+{
+  setlocale (LC_ALL, "");
+  bindtextdomain (GETTEXT_PACKAGE, LOCALEDIR);
+  bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
+  textdomain (GETTEXT_PACKAGE);
+
+  application = gtk_application_new ("org.gnome.Tali", 0);
+  g_signal_connect (application, "startup", G_CALLBACK (GyahtzeeStartup), NULL);
+  g_signal_connect (application, "activate", G_CALLBACK (GyahtzeeActivateGame), NULL);
+  g_signal_connect (application, "handle-local-options", G_CALLBACK (GyahtzeeLocalOptions), NULL);
+
+  /* Reset all yahtzee variables before parsing args */
+  YahtzeeInit ();
+
+  g_application_add_main_option_entries (G_APPLICATION (application), yahtzee_options);
+
+  settings = g_settings_new ("org.gnome.Tali");
+
+  g_set_application_name (_(appName));
+  g_set_prgname ("org.gnome.Tali");
   gtk_window_set_default_icon_name ("org.gnome.Tali");
 
   g_application_run (G_APPLICATION (application), argc, argv);
-- 
GitLab

